File: facade-pattern/home-theatre/home-theatre.ts
// Sub-systems
class Amplifier {
  on(): void {
    console.log("Amplifier is now ON.");
  }

  off(): void {
    console.log("Amplifier is now OFF.");
  }
}

class Projector {
  on(): void {
    console.log("Projector is now ON.");
  }

  off(): void {
    console.log("Projector is now OFF.");
  }
}

class Lights {
  dim(): void {
    console.log("Lights are dimmed for the movie.");
  }

  brighten(): void {
    console.log("Lights are back to full brightness.");
  }
}

class TVScreen {
  down(): void {
    console.log("Screen is lowered.");
  }

  up(): void {
    console.log("Screen is raised.");
  }
}

// Facade
class HomeTheatre {
  private amplifier: Amplifier;
  private projector: Projector;
  private lights: Lights;
  private screen: TVScreen;

  constructor() {
    this.amplifier = new Amplifier();
    this.projector = new Projector();
    this.lights = new Lights();
    this.screen = new TVScreen();
  }

  watchMovie(): void {
    console.log("Preparing to watch a movie now......");
    this.amplifier.on();
    this.projector.on();
    this.lights.dim();
    this.screen.down();
    console.log("Movie is ready to play!!!!!");
  }

  endMovie(): void {
    console.log("Shutting down movie now.");
    this.amplifier.off();
    this.projector.off();
    this.lights.brighten();
    this.screen.up();
    console.log("Movie shutdown completed.");
  }
}

const theatre = new HomeTheatre();
theatre.watchMovie();

console.log("\n \nTrying to end.....");
theatre.endMovie();

File: facade-pattern/home-theatre/home-theatre.rs
// Subsystems
struct Amplifier;

impl Amplifier {
    fn on(&self) {
        println!("Amplifier is on");
    }

    fn off(&self) {
        println!("Amplifier is off");
    }
}

struct Projector;

impl Projector {
    fn on(&self) {
        println!("Projector is on");
    }

    fn off(&self) {
        println!("Projector is off");
    }
}

struct Lights;

impl Lights {
    fn dim(&self) {
        println!("Lights are dimmed for the movie.");
    }

    fn brighten(&self) {
        println!("Lights are back to full brightness.");
    }
}

struct Screen;

impl Screen {
    fn down(&self) {
        println!("Screen is lowered.");
    }

    fn up(&self) {
        println!("Screen is raised.");
    }
}

// Facade
struct HomeTheater {
    amplifier: Amplifier,
    projector: Projector,
    lights: Lights,
    screen: Screen,
}

impl HomeTheater {
    fn new() -> Self {
        HomeTheater {
            amplifier: Amplifier,
            projector: Projector,
            lights: Lights,
            screen: Screen,
        }
    }

    fn watch_movie(&self) {
        println!("Preparing to watch a movie now......");
        self.amplifier.on();
        self.projector.on();
        self.lights.dim();
        self.screen.down();
        println!("Movie is ready to play!!!!!");
    }

    fn end_movie(&self) {
        println!("Ending the movie now......");
        self.amplifier.off();
        self.projector.off();
        self.lights.brighten();
        self.screen.up();
        println!("Movie has ended!");
    }
}

fn main() {
    let theatre = HomeTheater::new();
    theatre.watch_movie();
    println!("\n\n--- Movie Finished ---\n");
    theatre.end_movie();
}
